### PR TITLE
Add lab roadmap-week05

### PR DESCRIPTION
**Pull request by UCLWilson**

### CODE REVIEW

**Comment by jonjoncardoso** [comment]
Closes #55 (once merged)

**Review by jonjoncardoso — CHANGES_REQUESTED** [review]
You pushed many files that should not be here.

Which IDE are you using? If VSCode, use the instructions I wrote on the README at this project [here](https://github.com/LSE-DSI/lse-ds202-course-notes#how-to-build-the-course-website).

Please delete all the html/css, etc files generated by Quarto and keep only the `.qmd`

File: weeks/week05/lab.qmd [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+1. Create a function as `boot.fn()`.
+
+    ```r
+    > boot.fn <- function(data, index)
+    + coef(lm(mpg ∼ horsepower, data = data, subset = index)) 
+    
+    > boot.fn(Auto, 1:392)
+    [1](Intercept)  horsepower 
+        39.9358610  -0.1578447
+    ```
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:285]: I think adding two different ways of coding the same function is confusion. Can we delete the first option? Generally when we create functions we use the curly brackets.

File: weeks/week05/lab.qmd [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+        call    3    -none- call   
+        K       1    -none- numeric
+        delta   2    -none- numeric
+        seed  626    -none- numeric
+	
+	```
+
+	Please be aware that, The `cv.glm()` function is part of the `boot` library. Meanwhile, you can explore the `cv.err` by yourself to see what is the meaning of 	`call`,`K`,`delta` and `seed`. This online webpage will be useful when interpreting the results. [Intepretation of crossvalidation result](https://stats.stackexchange.com/questions/48766/intepretation-of-crossvalidation-result-cv-glm)
+
+	🏋️ There is a bonus for the senior students. You can explore the package `caret` for the LOOCV estimation.
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:147]: I think it may not be the best idea to point to caret as the package is superceded by the tidymodels ecosystem.

File: weeks/week05/lab.qmd [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+	> cv.err$delta
+	[1]   1   1 
+	   24.23 24.23
+
+	# explore cv.err 
+	> summary(cv.err)
+
+    [1]	      Length Class  Mode   
+        call    3    -none- call   
+        K       1    -none- numeric
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:139]: I think we can achieve the same thing by instantiating an empty vector to store results. We don't have to specify the length.

File: weeks/week05/lab.qmd [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+	```
+
+	```r
+	> y<-rnorm(3) #try to replicate the result from x, but totally different
+	[1]：0.07050839 0.12928774 1.71506499
+	```
+
+	```r
+	> set.seed(123)
+	> y<-rnorm(3) #try to replicate the result from x again with the same seed
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:45]: I really think that the root mean squared error is a better statistic.

File: weeks/week05/lab.qmd [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+👉 [R for Data Science](https://r4ds.had.co.nz/index.html) (book by Hadley Wickham & Garrett Grolemund)
+</p>
+
+<details><summary> Step 1: The Validation Set Approach (15mins) </summary>
+<br>
+In this step, we will explore how to extract the subset of the whole dataset as a training dataset, and then estimate the test error rates of various linear models. The dataset is `Auto` from the `ISLR2` package.
+
+1. Split the `Auto` dataset into two halves, as randomly selecting 196 observations out of the original 392 observations.
+	
+	```{r}
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:25]: I have included rsample from the tidymodels ecosystem. It really simplifies things nicely, plus students have learned how to create train/test splits using base R in the last lesson.

File: weeks/week05/lab.qmd [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+	
+    ⚡️The `-train` index selects only the observations that are not in the training set.
+
+	```r
+	> attach(Auto)
+	> mpg
+	[1] You can see all the values from the 'mpg' column.
+	```
+	However, if you do not use `attach` to call the dataframe, you have to use the `$` function to have the same results.
+
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:72]: So many students asked me how to create a graph like the one below so it seemed like a good time to add this.

File: weeks/week05/lab.qmd [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+            Null deviance: 23819.0  on 391  degrees of freedom
+        Residual deviance:  9385.9  on 390  degrees of freedom
+        AIC: 2363.3
+
+        Number of Fisher Scoring iterations: 2
+	
+	# get the cross-validation results by the 'cv.glm' function
+	> cv.err <- cv.glm(Auto, glm.fit)
+	> cv.err$delta
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:130]: I would be wary of promoting caret as it has been superceded by tidymodels.

File: weeks/week05/lab.qmd [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+	> mpg
+	[1] You can see all the values from the 'mpg' column.
+	```
+	However, if you do not use `attach` to call the dataframe, you have to use the `$` function to have the same results.
+
+	```r
+	> Auto$mpg
+	[1] You can see all the values from the 'mpg' column.
+	```
+
```
Comment (sbramwell86) [inline weeks/week05/lab.qmd:77]: Can we delete Step 2? I think k-fold cross-validation and the bootstrap are more widely used resampling methods and the delta output for LOOCV is really confusing (what do the two values mean?). I don't think the website really adds much extra information to be honest. By contrast, if you specify K = 10, you get 10 MSEs which is more intuitive.

File: weeks/week05/lab.qmd [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+👉 [R for Data Science](https://r4ds.had.co.nz/index.html) (book by Hadley Wickham & Garrett Grolemund)
+</p>
+
+<details><summary> Step 1: The Validation Set Approach (15mins) </summary>
+<br>
+In this step, we will explore how to extract the subset of the whole dataset as a training dataset, and then estimate the test error rates of various linear models. The dataset is `Auto` from the `ISLR2` package.
+
+1. Split the `Auto` dataset into two halves, as randomly selecting 196 observations out of the original 392 observations. We performed splits using base R in the last lab. However, this can be done more easily using the `rsample` package. We first create a split object `Auto.split` using `prop = 0.5` to specify a 50/50 train/test split. We also specify `strata = mpg` as we want our train/test split to have a similar mean/standard deviation for `mpg`. We then create dataframes using the `training` and `testing` functions.
+	
+	```r
```
Comment (jonjoncardoso) [inline weeks/week05/lab.qmd:25]: (I'm making the changes myself as I review it, but I will leave some comments here for future recommendations)

Make this an R cell instead of simple syntax highligthing. That is, replace \`\`\`r for ```{r}`

File: weeks/week05/lab.qmd [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+from: markdown+emoji
+author: Xiaowei Gao & Stuart Bramwell
+---
+
+<p> Last week we learnt how to do the Classification Methods in R. As data scientists, we need to know how to examine the results we generated and also to justify our results.  So, we will focus on and practice the validation methods in the R language in this week's lab. 
+
+If you feel confident at the current stage, free to explore more on your own. We have provided you with some supplementary online resources : 
+
+<br>
+👉 [R for Data Analytics](https://rforanalytics.com/13.6-k-fold-cross-validation.html) (part of the courses by Abhay Singh)
```
Comment (jonjoncardoso) [inline weeks/week05/lab.qmd:14]: (I'm making the changes myself as I review it, but I will leave some comments here for future recommendations)

Use bullet points instead of <br>

File: weeks/week05/lab.qmd [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+	lm.fit <- lm(mpg ~ horsepower, data = Auto.train)
+	# estimate the 'mpg' values by the lm.fit model
+	lm.pred <- predict(lm.fit, Auto.test)
+	# calculate MSE
+	mean((Auto.test$mpg - lm.pred)^2)
+	[1] 25.59
+	```
+Therefore, we have estimated the test MSE for the linear regression model, **25.59**. (Well Done! 💪 )
+<br>
+Please note, here we use new functions, namely `lm` and `predict`. By typing `?`, you can find the official explanation of these functions.
```
Comment (jonjoncardoso) [inline weeks/week05/lab.qmd:52]: `lm` and `predict` are NOT NEW to students! They've been using it since W03

File: weeks/week05/lab.qmd (hunk 8 above)
Comment (jonjoncardoso) [reply weeks/week05/lab.qmd:62]: I am of two minds about this. I think I will cover only k-fold CV and bootstrap in the lecture, but the Chapter 6 of the book provides an interesting discussion of the difference between LOOCV and k-fold CV, in light of bias vs variance tradeoff.

File: weeks/week05/lab.qmd [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+
+</details>
+
+<details><summary> Step 3: k-Fold Cross-Validation (5mins)</summary>
+<br>
+It will be easy to follow the former procedure in Step 2, by the `cv.glm` to implement ***K-fold*** CV.
+
+1. Estimate CV errors corresponding to the polynomial fits of orders one to ten using ten-fold cross-validation (via `K = 10`). 
+
+	```{r}
```
Comment (jonjoncardoso) [inline weeks/week05/lab.qmd:191]: (I'm making the changes myself as I review it, but I will leave some comments here for future recommendations)

Make this an R cell instead of simple syntax highligthing. That is, replace ```r for ```{r}`

File: weeks/week05/lab.qmd [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+### Step 4.1 Estimating the Accuracy of a Linear Regression Model
+<br>
+In this step, we will use the boostrap approach to assess the variability of a coefficient estimate. For the sake of simplicity we will look at the relationship between `weight` and `horsepower` which appears to be linear.
+
+<br>
+
+1. Create a function as `boot.fn()` which 
+
+    ```r
+    boot.fn <- function(data, index) {
```
Comment (jonjoncardoso) [inline weeks/week05/lab.qmd:238]: (I'm making the changes myself as I review it, but I will leave some comments here for future recommendations)

Make this an R cell instead of simple syntax highligthing. That is, replace ```r for ```{r}`

File: weeks/week05/lab.qmd [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+        lm(horsepower ~ weight, data = data[index,])$coefficients
+
+    }  
+    ```
+
+`boot.fn` simply returns a vector of coefficient estimates. It takes two parameters: `data` and `index`. `data` is a placeholder for the dataframe used in the model. `index` is a placeholder for the sample used to subset the dataframe. Other than this, the body of the function should look familiar. We are estimating a linear model where we are looking to predict `horsepower` by `weight`, and then extracting the coefficients.
+
+2. Compare the results from bootstrap estimates and the standard estimates 
+    
+    ```r
```
Comment (jonjoncardoso) [inline weeks/week05/lab.qmd:249]: (I'm making the changes myself as I review it, but I will leave some comments here for future recommendations)

Make this an R cell instead of simple syntax highligthing. That is, replace ```r for ```{r}`

**Review by jonjoncardoso — CHANGES_REQUESTED** [review]

File: weeks/week05/lab.qmd (hunk 6 above)
Comment (jonjoncardoso) [reply weeks/week05/lab.qmd:57]: Good

**Comment by jonjoncardoso** [comment]
Merged by c03455828ae08077d3fdacf60f728246e053d20c
